pkg/util/lock.go

```diff
@@ -111,30 +114,167 @@ func (l *PgAdvisoryLock) TryLock() (bool, error) {
+	return l.Release()
+}
+
+// Locked returns if the instance was able to obtain the locks.
+func (l *PgLeaderLock) Locked() bool {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+	return l.obtained
+}
+
+// Release releases the already obtained locks.
+func (l *PgLeaderLock) Release() error {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	defer l.connCleanUp()
+	if !l.obtained {
+		return fmt.Errorf("can't release while not holding the lock")
 	}
+
+	defer func() { l.obtained = false }()
+
+	unlocked, err := l.unlock()
 	if err != nil {
-		return false, err
+		return err
+	}
+	if !unlocked {
+		log.Debug("msg", fmt.Sprintf("false release for group id %d", l.groupLockID))
+	}
+
+	return nil
+}
+
+type PgAdvisoryLock struct {
+	conn         *pgx.Conn
+	connStr      string
+	groupLockID  int64
+	afterConnect AfterConnectFunc
+
+	mutex sync.RWMutex
+}
+
+type AdvisoryLock interface {
+	GetAdvisoryLock() (bool, error)
+	GetSharedAdvisoryLock() (bool, error)
+	Unlock() (bool, error)
+	UnlockShared() (bool, error)
+	Close()
+}
+
+// PgAdvisoryLock is a AdvisoryLock
+var _ AdvisoryLock = (*PgAdvisoryLock)(nil)
+
+// NewPgAdvisoryLock creates a new instance with specified lock ID, connection pool and lock timeout.
+func NewPgAdvisoryLock(groupLockID int64, connStr string) (*PgAdvisoryLock, error) {
+	lock := &PgAdvisoryLock{
+		connStr:     connStr,
+		groupLockID: groupLockID,
+	}
+	return lock, nil
+}
+
+func (l *PgAdvisoryLock) getConn(connStr string, cur, maxRetries int) (*pgx.Conn, error) {
+	if maxRetries == cur {
+		return nil, fmt.Errorf("max attempts reached. giving up on getting a db connection")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), waitForConnectionTimeout)
+	defer cancel()
+
+	cfg, err := pgx.ParseConfig(connStr)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing config connection: %w", err)
+	}
+
+	lockConn, err := pgx.ConnectConfig(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("error getting DB connection: %w", err)
 	}
-	defer func() {
+
+	if l.afterConnect != nil {
+		err = l.afterConnect(ctx, lockConn)
 		if err != nil {
-			l.connCleanUp()
+			log.Error("msg", "Invalid connection", "err", err)
+			return l.getConn(connStr, cur+1, maxRetries)
 		}
-	}()
-	rows, err := l.conn.Query(context.Background(), "SELECT pg_try_advisory_lock($1)", l.groupLockID)
+	}
+
+	return lockConn, nil
+}
+
+func (l *PgAdvisoryLock) GetAdvisoryLock() (bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	return l.getAdvisoryLock()
+}
+
+func (l *PgAdvisoryLock) getAdvisoryLock() (bool, error) {
+	return l.runLockFuncion("SELECT pg_try_advisory_lock($1)")
+}
+
+func (l *PgAdvisoryLock) GetSharedAdvisoryLock() (bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	return l.getSharedAdvisoryLock()
+}
+
+func (l *PgAdvisoryLock) getSharedAdvisoryLock() (bool, error) {
+	return l.runLockFuncion("SELECT pg_try_advisory_lock_shared($1)")
+}
+
+func (l *PgAdvisoryLock) runLockFuncion(query string) (bool, error) {
```

Comment:
```suggestion
func (l *PgAdvisoryLock) runLockFunction(query string) (bool, error) {
```

Reply:
it's sad that even with the diff in front of me it took a good minute for me to figure out what changed 😞